File: public/js/api/nodetype.js
import { protected_fetch } from "./util.js";

export const get_all = () => protected_fetch(`api/nodetype`);

export const get_by_id = id => protected_fetch(`api/nodetype/${id}`);

export const add = nt => protected_fetch(`api/nodetype`, {
    method: "POST",
    body: JSON.stringify(nt)
});

export const update = nt => protected_fetch(`api/nodetype/${nt.id}`, {
    method: "POST",
    body: JSON.stringify(nt)
});

export const remove = nt => protected_fetch(`api/nodetype/${nt.id}`, {
    method: "DELETE"
});

File: web/node_type.go
package web

import (
	"encoding/json"
	"mt-hosting-manager/types"
	"net/http"

	"github.com/gorilla/mux"
)

func (a *Api) GetNodeTypes(w http.ResponseWriter, r *http.Request) {
	list, err := a.repos.NodeTypeRepo.GetAll()
	Send(w, list, err)
}

func (a *Api) GetNodeType(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	nt, err := a.repos.NodeTypeRepo.GetByID(vars["id"])
	Send(w, nt, err)
}

func (a *Api) CreateNodeType(w http.ResponseWriter, r *http.Request, c *types.Claims) {
	nt := &types.NodeType{}
	err := json.NewDecoder(r.Body).Decode(nt)
	if err != nil {
		SendError(w, 500, err)
		return
	}
	err = a.repos.NodeTypeRepo.Insert(nt)
	Send(w, nt, err)
}

func (a *Api) UpdateNodeType(w http.ResponseWriter, r *http.Request, c *types.Claims) {
	nt := &types.NodeType{}
	err := json.NewDecoder(r.Body).Decode(nt)
	if err != nil {
		SendError(w, 500, err)
		return
	}
	err = a.repos.NodeTypeRepo.Update(nt)
	Send(w, nt, err)
}

func (a *Api) DeleteNodeType(w http.ResponseWriter, r *http.Request, c *types.Claims) {
	vars := mux.Vars(r)
	err := a.repos.NodeTypeRepo.Delete(vars["id"])
	Send(w, true, err)
}
